### ISSUE TITLE
Items with custom entities are deleted on toss.

### ISSUE THREAD

**Issue by Thutmose** [SpongePowered/SpongeForge#1810]
**I am currently running...**

- SpongeForge version: 1.11.2-2476-6.1.0-BETA-2699
- Forge version: 13.20.1.2510
- Java version: 1.8.0_111
- Operating System: Windows 10

 - [ ] Linux (Ubuntu, Debian, Fedora, etc)
 - [ ] macOS
 - [X] Windows

- Plugins/Mods: 
Minecraft, Minecraft Coder Pack, Forge Mod Loader, Minecraft Forge, SpongeAPI, SpongeForge, Thut Core, In-game NBTEdit - Pokecube Edition, Pokécube Core, Pokecube Model Loader, Thut's Wearables, Thut's Bling, Pokécube Revival, Pokécube Mobs, Pokecube Movess, Pokecube Origin, Pokecube Compat, Pokecube Mystery Dungeon, Thut's Tech

**What is the expected result?**
When Q is pressed to toss an item, it should toss (or have whatever the plugins do affect the toss), regardless of whether it should produce and EntityItem or its own custom entity type.

**What is the current result?**
When Tossed, non-EntityItem entities will be deleted due to spongeforge trying to cast the entity to EntityItem

Here is the relevant section of log:
https://gist.github.com/Thutmose/1a1d66d6ec925ea8e66faa0216389413

Here is the relvant section of Pokecube where it makes the item use a custom entity instead of an EntityItem:
https://github.com/Thutmose/Pokecube/blob/1.10/Pokecube%20Core/src/main/java/pokecube/core/items/pokecubes/Pokecube.java#L396-L424

Edit: For completeness, confirmed that this occurs with "latest" versions of forge and spongeforge:
https://gist.github.com/Thutmose/0204ae3f1060eed1253ccec221cebad2

**Comment by ryantheleach** [comment]
Should Pokecube be extending EntityItem?

**Comment by Thutmose** [comment]
not nessisarily, the method allows returning any "Entity" as the item when tossed, something could say spawn an exp orb when tossed instead, which would then be an EntityXPOrb, which does not extend EntityItem.

**Comment by gabizou** [comment]
Looks like an error on our end. We're expecting that all entities dropped from the player interactions are always `EntityItem`s. It never occurred to us that there were alternative options for when a player drops an item. Should see a fix soon for both 1.11 and 1.12.

**Comment by Keaton1188** [comment]
Any update on that fix?? Also having a very similar issue.. When pressing Q the item doesn't drop it just stays in your inv, this doesn't happen all the time but multiple people have reported it to me. When forcing the item out your inv, so pressing E the. Picking it up and throwing it out it will just delete this item and not drop..1.12.2 this is happening on.

**Comment by smmmadden** [comment]
any update on a fix for this?  I'm seeing it in build 2528 and 2538 for 1.12.2.

**Comment by Thutmose** [comment]
As of my testing in 2831, this issue has been partially addressed, but not completely.

It seems that the ItemTossEvent is not being forwarded to forge mods when the tossed item is a non-EntityItem

**Comment by Zidane** [comment]
```java
    public ItemTossEvent(EntityItem entityItem, EntityPlayer player)
```

Signature literally requires it to be an EntityItem and that is being called in SpongeForge.

**Comment by Thutmose** [comment]
When I get home, I will look into what forge does then, as forge sends an itemtossevent in this case, I suspect it is making a toss event for the conventional itemstack, then making the new entity and actually tossing that

**Comment by Thutmose** [comment]
So, it seems that forge converts the itemstack from an item to an custom entity at the point when it is added to the world, so during all of the cases where ItemTossEvent is used, it is still an itemstack.

https://github.com/MinecraftForge/MinecraftForge/blob/1.12.x/src/main/java/net/minecraftforge/common/ForgeInternalHandler.java#L51-L64

**Comment by Thutmose** [comment]
I can confirm that this bug still occurs with spongeforge 1.12.2-2655-7.1.0-BETA-3037

Where do you intercept the dropping of the item? is it before or after it is added to the world?

**Comment by gabizou** [comment]
Sponge intercepts item drops from various places, but always before the entity item is created, and when the item entity is spawned (usually captured for later spawning with proper bulk events). This issue *might* be resolved once I do [finish ~~rewriting~~ the drop capture management](https://github.com/SpongePowered/SpongeCommon/compare/stable-7...refactor/entity-drop-compatibility).

**Comment by Thutmose** [comment]
I am also working on making a tiny test mod for this case, it will add a single item that when dropped, spawns a chicken.

hopefully it will help with narrowing things down.

**Comment by Thutmose** [comment]
[eggtoss.zip](https://github.com/SpongePowered/SpongeForge/files/1929894/eggtoss.zip)
the zip contains the obfuscated, deobfuscated and sources for the mod

Here is a mod that displays the current bug, it adds a single item which can be found in the "tools" creative tab.

Without spongeforge, dropping this item spawns a chicken at the player's location.

With spongeforge, dropping the item does nothing

**Comment by gabizou** [comment]
@Thutmose thanks for the update, I think I stumbled on the root of the issue, Forge handles creating the custom entity through an internal Forge listener (which makes sense why it may not be working now). I'll update the issue as I progress with the refactoring).

**Comment by Aaron1011** [comment]
@gabizou: We took over how Forge handles custom item entities in commit https://github.com/SpongePowered/SpongeForge/commit/ff5615eb4b442c95b8a01187305a2710dd52a982#diff-fd74c12711b592b4abd8fd1f7d34f37a. However, that was several years ago, and Forge's behavior may have changed since then.

**Comment by gabizou** [comment]
Yes, I've already started writing the implementation to support custom item entities. It'll show up in #2110 as soon as I've completed that rewrite.

**Comment by gabizou** [comment]
@Thutmose I've gotten [some centralization of code](https://github.com/SpongePowered/SpongeCommon/commit/fc3df57d81f0d8616f8811e43bdbfa3b622c7179#diff-88b8beb17c23098a0b640f4891177c72R509) to handle entity spawns and when they are items, to check for their custom entities to spawn (this is wrapped around in our "common" codebase to allow us to target vanilla, while still [overwriting as necessary to retain forge compatibility](https://github.com/SpongePowered/SpongeForge/commit/2fdfe1f13502b7c3236c3556115c3a7dc48193c6#diff-2455e8bda022a4836c472d6e204fe3ffR68)

I hope this should work, and I'll use the test mod in a test plugin of sorts for SpongeForge to verify this works as expected.

**Comment by Thutmose** [comment]
Hopefully it goes well, the test mod was the simplest case for displaying this particular bug, once it is solved, I can probably start looking for other bugs between pokecube and spongeforge, this was the last one I was made aware of, if I find more, I will either try to fix on my end, or will see about making more simple test mods that display the behaviour

**Comment by gabizou** [comment]
Verified now that custom item drops are fully functional in all aspects throughout Sponge's drop handling and entity spawns (no room for any entity to be spawned into the world unless someone does the actual duplicate code to spawn an entity without calling the spawn entity method).

The issue will be closed once #2110 is merged.

**Comment by Thutmose** [comment]
I can confirm that this issue no longer occurs in spongeforge 3074
